src/utils/db/car_stream.rs

```diff
@@ -137,6 +140,68 @@ impl<ReaderT: AsyncBufRead> Stream for CarStream<ReaderT> {
     }
 }
 
+pin_project! {
+    pub struct CarWriter<W> {
+        #[pin]
+        inner: W,
+        buffer: Vec<u8>,
+    }
+}
+
+impl<W: AsyncWrite> CarWriter<W> {
+    pub fn new_carv1(roots: Vec<Cid>, writer: W) -> anyhow::Result<Self> {
+        let car_header = CarHeader { roots, version: 1 };
+
+        let mut header_uvi_frame = BytesMut::new();
+        UviBytes::default().encode(Bytes::from(to_vec(&car_header)?), &mut header_uvi_frame)?;
+
+        Ok(Self {
+            inner: writer,
+            buffer: header_uvi_frame.to_vec(),
+        })
+    }
+}
+
+impl<W: AsyncWrite> Sink<(Cid, Vec<u8>)> for CarWriter<W> {
+    type Error = io::Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let this = self.project();
+        if !this.buffer.is_empty() {
+            let bytes_written = ready!(this.inner.poll_write(cx, &this.buffer))?;
+            *this.buffer = this.buffer[bytes_written..].to_vec();
+            if !this.buffer.is_empty() {
+                return Poll::Pending;
+            }
+        }
+        Poll::Ready(Ok(()))
+    }
+    fn start_send(self: Pin<&mut Self>, item: (Cid, Vec<u8>)) -> Result<(), Self::Error> {
+        let this = self.project();
+
+        let payload = [item.0.to_bytes(), item.1].concat();
+        let mut var_bytes = [0; 10];
+        let len = payload.len().encode_var(&mut var_bytes);
```

Comment:
We already have code for encoding a block.

```rust
let block = ...;
let buffer = ...;
block.write(&mut buffer)?;
```

Reply:
Is there some support for `impl std::io::Write` for `BytesMut`?

Reply:
Hmm, maybe not. Then write to an intermediate vector.

Reply:
Maybe `bytes::buf::Writer` can do the trick.